addons/binding/org.openhab.binding.kostalinverter/src/main/java/org/openhab/binding/internal/kostal/inverter/thirdgeneration/ThirdGenerationHandler.java

```diff
@@ -0,0 +1,493 @@
+import org.eclipse.smarthome.core.library.unit.SIUnits;
+import org.eclipse.smarthome.core.library.unit.SmartHomeUnits;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonObject;
+
+/**
+ * The {@link ThirdGenerationHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author René Stakemeier - Initial contribution
+ */
+@NonNullByDefault
+public class ThirdGenerationHandler extends BaseThingHandler {
+
+    /*
+     * operations used for authentication
+     */
+    private static final String AUTH_START = "/auth/start";
+    private static final String AUTH_FINISH = "/auth/finish";
+    private static final String AUTH_CREATE_SESSION = "/auth/create_session";
+
+    /*
+     * operations used for gathering process data from the device
+     */
+    private static final String PROCESSDATA = "/processdata";
+
+    /*
+     * After the authentication the result (the session id) is stored here and used to "sign" future requests
+     */
+    private @Nullable String sessionId;
+    /*
+     * The configuration file containing the host, the password and the refresh interval
+     */
+    private @NonNullByDefault({}) ThirdGenerationConfiguration config;
+
+    private @Nullable ScheduledFuture<?> refreshScheduler;
+
+    private @Nullable HttpClient httpClient;
+
+    private ThirdGenerationInverterTypes inverterType;
+
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+    /**
+     * Constructor of this class
+     *
+     * @param thing        the thing
+     * @param httpClient   the httpClient used for communication
+     * @param inverterType the type of the device
+     */
+    public ThirdGenerationHandler(Thing thing, HttpClient httpClient, ThirdGenerationInverterTypes inverterType) {
+        super(thing);
+        this.inverterType = inverterType;
+        this.httpClient = httpClient;
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        // All channels are readonly and updated by the scheduler
+    }
+
+    @Override
+    public void dispose() {
+        if (refreshScheduler != null) {
+            refreshScheduler.cancel(true);
+			refreshScheduler = null;
+        }
+        super.dispose();
+    }
+
+    @Override
+    public void initialize() {
+        config = getConfigAs(ThirdGenerationConfiguration.class);
+        // temporary value while initializing
+        updateStatus(ThingStatus.UNKNOWN);
+
+        // Start the authentication
+        scheduler.schedule(this::authenticate, 1, TimeUnit.SECONDS);
```

Comment:
Theoretically the authenticate schedule call can conflict with the scheduled update call. Just something to consider if you need some extra safe programming to handle such a case.

Reply:
Calling the authenticate at first is the safe-programming-part :-)

If the call for value fails with http error (503 if I recall right), the authenticate is triggered anyway.
Calling the authenticate at first will prevent the error on the first communication (assumption the device is actualy configured right).